examples/dao/assets/dao-app-approval.py

```diff
@@ -0,0 +1,449 @@
+import sys
+sys.path.insert(0,'..')
+
+from algobpy.parse import parse_params
+from pyteal import *
+
+def approval_program():
+    """
+    A stateful app which will encode governance rules. Saves
+    deposit, min_support, min_duration, max_duration, url.
+
+    Commands:
+        add_proposal            Save proposal record in lsig
+        deposit_vote            records deposited votes in voter.account
+        register_vote           register user votes in proposal_lsig
+        execute                 executes a proposal
+        withdraw_vote_deposit   unlock the deposit and withdraw tokens back to the user
+        clear_vote_record       clears Sender local state by removing a record of vote cast from a not active proposal
+        clear_proposal          clears proposal record and returns back the deposit
+    """
+
+    # Verfies deposit of gov_token to an address for a transaction. Args:
+    # * tx_index (index of deposit transaction)
+    # * receiver (receiver of gov_token)
+    def verify_deposit(tx_index: Int, receiver: Addr):
+        return Assert(And(
+            Global.group_size() >= tx_index,
+            Gtxn[tx_index].type_enum() == TxnType.AssetTransfer,
+            Gtxn[tx_index].xfer_asset() == Tmpl.Int("TMPL_GOV_TOKEN"),
+            Gtxn[tx_index].asset_receiver() == receiver,
+            Gtxn[tx_index].asset_amount() >= Int(0)
+        ))
+
+    # scratch vars to save result() & is_proposal_active()
+    scratchvar_result = ScratchVar(TealType.uint64)
+    scratchvar_is_proposal_active = ScratchVar(TealType.uint64)
+
+    # Fetch key(s) from Txn.Accounts[1] (passed as proposalLsig), for current app
+    # these are used throughout the application
+    voting_start = App.localGet(Int(1), Bytes("voting_start"))
+    voting_end = App.localGet(Int(1), Bytes("voting_end"))
+    proposal_id = App.localGet(Int(1), Bytes("id"))
+    yes_count = App.localGet(Int(1), Bytes("yes"))
+    no_count = App.localGet(Int(1), Bytes("no"))
+    execute_before = App.localGet(Int(1), Bytes("execute_before"))
+    executed = App.localGet(Int(1), Bytes("executed"))
+    proposal_type = App.localGet(Int(1), Bytes("type"))
+    proposal_from = App.localGet(Int(1), Bytes("from"))
+    asa_id = App.localGet(Int(1), Bytes("asa_id"))
+    recipient = App.localGet(Int(1), Bytes("recipient"))
+    amount = App.localGet(Int(1), Bytes("amount"))
+
+    # Computes result of proposal. Saves result in scratch. Args:
+    # * idx - index of account where proposal_lsig.address() is passed
+    # NOTE: idx == Int(0) means proposalLsig is Txn.sender()
+    def compute_result(idx: Int):
+        return Seq([
+            Cond(
```

Comment:
i thought so, but here we're computing a result against latest time (`Global.latest_timestamp()`). So result could be different for different block timestamps.

Reply:
As far as I know scratch space is reset after the execution. Could you validate?

Reply:
sorry, i don't follow.
+ we're only using this function once per branch (for each execution), so it will only be filled once (you're right about scratch space is reset after the execution).
+ but even if this function was called multiple times (at a single branch), then we should compute result again (yes, scratch space will be filled, but we have moven forward in time, so result could be different).

Am i saying something wrong here?

Reply:
How possible that the result may change within the same transaction?

Reply:
ah, yes. OK, i was trying to do it, i keep getting a python error (with slot assignment):
![image](https://user-images.githubusercontent.com/33264364/134501354-2eebedeb-43d6-4d02-8e9d-ef58a44ddd04.png)
![Screenshot from 2021-09-23 17-13-27](https://user-images.githubusercontent.com/33264364/134501404-75ba0e3c-7f13-47f7-aeab-2f12556c0c8e.png)